spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractMessageListenerContainer.java

```diff
@@ -704,6 +711,13 @@ protected void publishContainerStoppedEvent() {
 		}
 	}
 
+	protected void publishConcurrentContainerStoppedEvent() {
```

Comment:
This is not correct.
It has to be only in the `ConcurrentMessageListenerContainer`.
More over I think that event has to be a part of logic of the overridden `publishContainerStoppedEvent()`.

Reply:
As per the earlier discussion, you have suggested not to make any changes related to the 'ContainerStoppedEvent'. So, I have added new method to publish 'ConcurrentContainerStoppedEvent'. I can move code 'publishConcurrentContainerStoppedEvent()' to ConcurrentMessageListenerContainer. But, we can not have same code to publish both 'ContainerStoppedEvent'  and 'ConcurrentMessageListenerContainer'.

Reply:
That's not correct.
To not break the current behavior we still have to emit both events from the `ConcurrentMessageListenerContainer`.
And that new even is really just a responsibility of the `ConcurrentMessageListenerContainer`: don't mix concerns of the encapsulation.

We might just don't need a new method, but override that `publishContainerStoppedEvent()` in the `ConcurrentMessageListenerContainer` to publish `ConcurrentContainerStoppedEvent` as well.

Reply:
Please find the following points.

1) I have not made any code changes that will alter the 'ContainerStoppedEvent'.
	
2) In single method we cannot publish both 'ContainerStoppedEvent' and 'ConcurrentContainerStoppedEvent'. **'ConcurrentContainerStoppedEvent' has to be published from the 'childStopped' method in 'org.springframework.kafka.listener.ConcurrentMessageListenerContainer' once all the Containers are stopped.** We cannot publish both.

3) Following events are published from their own methods in KafkaMessageListenerContainer.   
   ConsumerPartitionResumedEvent, ConsumerStoppingEvent, ConsumerStoppedEvent etc....

In similar way, i have written a new method 'publishConcurrentContainerStoppedEvent' in 'org.springframework.kafka.listener.AbstractMessageListenerContainer'. Please let me know if you want me to move this method to 'org.springframework.kafka.listener.ConcurrentMessageListenerContainer'.

Please give your idea.


**Apart from that, i will remove all the changes that are not related to publishing 'ConcurrentContainerStoppedEvent'.**

Reply:
Right. As I said: the `ConcurrentContainerStoppedEvent` is a responsibility only of the `ConcurrentMessageListenerContainer`.
Therefore the method to emit (and respective logic) must be only in the `ConcurrentMessageListenerContainer`.
I'm not sure why you still insist keeping it in the `AbstractMessageListenerContainer`.
This is just wrong by the logic.

Well, if this new event must be emitted from another place, then the method override might not be the case.
So, no need to do any changes to the `publishContainerStoppedEvent()`, but just emit `ConcurrentContainerStoppedEvent` from where you think is appropriate.
However I still don't believe that the place has to be `AbstractMessageListenerContainer`...